### PR TITLE
Fix: UnsupportedOperationException when calling request.getBody on no…

### LINKED ISSUE #644: "java.lang.UnsupportedOperationException: direct buffer" when calling request.getBody
**To Reproduce**

Calling `request.getBody` when the request body is non-empty ends up with an exception.

Run:

```scala
import zio._
import zhttp.http._
import zhttp.service.Server

object ZioHttpTest extends App {
  val app = Http.collectM[Request] { case r =>
    r.getBody.map(_ => Response.text("Hello World!"))
  }

  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] =
    Server.start(8090, app).exitCode
}
```

And invoke:

```
curl -X POST -d "abc" http://localhost:8090
```

I'm getting:

```
<html>
  <head>
  </head>
  <body>
   <h1>Internal Server Error</h1>
   <pre><div>java.lang.UnsupportedOperationException: direct buffer</div><div>	at io.netty.buffer.PooledUnsafeDirectByteBuf.array(PooledUnsafeDirectByteBuf.java:226)</div><div>	at io.netty.buffer.CompositeByteBuf.array(CompositeByteBuf.java:793)</div><div>	at zhttp.http.Request.$anonfun$getBody$2(Request.scala:25)</div><div>	at zio.internal.FiberContext.evaluateNow(FiberContext.scala:490)</div><div>	at zio.internal.FiberContext.$anonfun$fork$17(FiberContext.scala:770)</div><div>	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)</div><div>	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)</div><div>	at java.base/java.lang.Thread.run(Thread.java:832)</div></pre>
  </body>
</html>
```

**Desktop (please complete the following information):**

This came up when migrating tapir to use latest zhttp in https://github.com/softwaremill/tapir/pull/1686 (cc @guizmaii)

```
"io.d11" %% "zhttp" % "1.0.0.0-RC18"
```

### PR DESCRIPTION
**Pull request by sumawa**
To resolve #644

### CODE REVIEW

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 436653.96**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 437020.73**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 1]
```diff
@@ -133,7 +133,13 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
         assertM(res)(equalTo(string.length.toString))
       }
-    }
+    } +
+      testM("POST Echo request body back") {
+        val path    = !! / "echoRequestBody"
+        val bodyMsg = "Hello World!"
+        val res     = request(path, Method.POST, bodyMsg).flatMap(_.getBodyAsString)
+        assertM(res)(equalTo(bodyMsg))
+      }
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala:141]: why is this test added?

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 435081.58**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala (hunk 1 above)
Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala:141]: In general, when we see our test cases, there are very few tests added for Methods like POST/PUT/DELETE also there are fewer examples with non GET methods.

Comment (tusharmath) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala:141]: Try to only have a max of 3 lines in any test.

Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala:141]: I wrote this specific test to simulate and fail the request.getBody

File: zio-http/src/main/scala/zhttp/http/Request.scala [hunk 2]
```diff
@@ -22,7 +22,11 @@ trait Request extends HeaderExtension[Request] { self =>
   /**
    * Decodes the content of request as a Chunk of Bytes
    */
-  def getBody: Task[Chunk[Byte]] = getBodyAsByteBuf.flatMap(buf => Task(Chunk.fromArray(buf.array())))
+  def getBody: Task[Chunk[Byte]] = getBodyAsByteBuf.flatMap(buf =>
+    Task(
+      Chunk.fromArray(ByteBufUtil.getBytes(buf)),
+    ),
+  )
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Request.scala]: (nitpick) keep everything in one line

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 433370.87**

File: zio-http/src/main/scala/zhttp/http/Request.scala (hunk 2 above)
Comment (sumawa) [reply zio-http/src/main/scala/zhttp/http/Request.scala]: Done

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 433205.57**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala (hunk 1 above)
Comment (tusharmath) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala:141]: Check the test written for `echo`

Comment (tusharmath) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala:141]: Test for StaticApp are defined under `def staticAppSpec`.

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 3]
```diff
@@ -169,9 +174,13 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
-    case Method.GET -> !! / "success"       => ZIO.succeed(Response.ok)
-    case Method.GET -> !! / "failure"       => ZIO.fail(new RuntimeException("FAILURE"))
-    case Method.GET -> !! / "get%2Fsuccess" => ZIO.succeed(Response.ok)
+    case Method.GET -> !! / "success"                => ZIO.succeed(Response.ok)
+    case Method.GET -> !! / "failure"                => ZIO.fail(new RuntimeException("FAILURE"))
+    case Method.GET -> !! / "get%2Fsuccess"          => ZIO.succeed(Response.ok)
+    case req @ Method.POST -> !! / "echoRequestBody" =>
+      for {
+        chunk <- req.getBody
+      } yield Response.text(chunk.map(_.toChar).mkString(""))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: Avoid writing tests on `staticApp`. They are hard to maintain and not co-located

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 435653.32**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 443500.71**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala (hunk 3 above)
Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: Done, moved to DynamicAppSpec, but under separate suite for testing getBody, not to mess with the existing tests

Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: Later in a separate PR when I will add support for setBody, I will try to organise tests under same suite for all Request APIs

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 4]
```diff
@@ -89,6 +89,17 @@ object ServerSpec extends HttpRunnableSpec(8088) {
           val res = app.requestBodyAsString()
           assertM(res)(equalTo("abc"))
         }
+      } + suite("Echo With getBody") {
+        val app = Http.collectM[Request] { case req @ Method.POST -> !! / "echoRequestBody" =>
+          for {
+            chunk <- req.getBody
+          } yield Response.text(chunk.map(_.toChar).mkString(""))
+        }
+        testM("POST Echo request body back") {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: Move the test to `def requestSpec`. You don't need such as complex application to test this failure.

Something like this should also produce the same error (You will need to check)
```scala
Http.collectM[Request] { case req => req.getBody.as(Response.ok) }
```

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 443131.84**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 445206.90**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala (hunk 4 above)
Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: Done

Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: I am still using POST and content

To reproduce the bug, only POST request with some content fails
val res = app.requestStatus(!!, Method.POST, "some text").    // fails

GET request passes 
val res = app.requestStatus()  /// this succeeds 

That is why I am keeping it POST, so as to keep close to the issue raised

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala (hunk 1 above)
Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala:141]: Done

**Review by tusharmath — APPROVED** [review]

**Pull request merged**